src/utils/daos_hdlr.c

```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+				}
+				rc = daos_cont_open(ap->dst_pool, ap->dst_c_uuid, DAOS_COO_RW,
+					&ap->dst_cont, dst_cont_info, NULL);
+				rc = dfs_mount(ap->dst_pool, ap->dst_cont, O_RDWR, &(daos_dst_file->dfs));
+				if (rc) {
+					fprintf(stderr, "dfs mount on destination failed: %s, %d\n",
+						ap->dst_path, rc);
+					D_GOTO(out, rc);
+				}
+			} else {
+				daos_dst_file->type = POSIX;
+			}
+		}
+	}
+	/* create DAOS POSIX destination container if it doesn't exist */
+	if (daos_dst_file->type == DAOS) {
+		if (daos_uuid_valid(ap->src_c_uuid) && !daos_uuid_valid(ap->dst_c_uuid)) {
+			/* create POSIX destination container if it doesn't exist */
+	    		if (uuid_is_null(ap->c_uuid))
+				uuid_generate(ap->dst_c_uuid);
+			rc = dfs_cont_create(ap->dst_pool, ap->dst_c_uuid, NULL, &(ap->dst_cont),
+						&(daos_dst_file->dfs));
+			if (rc != 0) {
+				fprintf(stderr, "failed to create container: %s (%d)\n",
+					d_errdesc(rc), rc);
+				D_GOTO(out, rc);
+			}
+			fprintf(stdout, "Successfully created container "DF_UUIDF"\n",
+				DP_UUID(ap->dst_c_uuid));
+		}
+	}
+	/* TODO: make sure to cleanup on errors, such as if the source pool and
+	 * container are connected, but the dfs_mount fails */
+out:
+	return rc;
+}
+
+static void
+daos_file_set_defaults(daos_file_t *daos_file)
+{
+	/* set defaults for daos_file struct */
+	daos_file->type	  = DAOS;
+	daos_file->fd     = -1; 
+	daos_file->offset = 0; 
+	daos_file->obj    = NULL; 
+	daos_file->dfs    = NULL; 
+}
+
+int
+fs_copy_hdlr(struct cmd_args_s *ap)
+{
+	/* TODO: add check to make sure all required arguments are 
+	 * provided */
+	int rc = 0;
+	daos_cont_info_t src_cont_info;
+	daos_cont_info_t dst_cont_info;
+	struct duns_attr_t dst_dattr = {0};
+	struct duns_attr_t src_dattr = {0};
+
+	/* daos_file_t abstraction for POSIX and DFS files */
+	daos_file_t daos_src_file;
+	daos_file_t daos_dst_file;
+	daos_file_set_defaults(&daos_src_file);
+	daos_file_set_defaults(&daos_dst_file);
+	fs_copy_connect(&daos_src_file, &daos_dst_file, ap,
+			   &src_cont_info, &dst_cont_info,
+			   &src_dattr, &dst_dattr);
+	/* set paths based on file type for source and destination */
+	char *name = NULL, *dname = NULL;
+	if (daos_src_file.type == POSIX && daos_dst_file.type == DAOS) {
+		ap->dst_path = "/";
+		parse_filename(ap->src_path, &name, &dname);
+		dfs_set_prefix(daos_dst_file.dfs, dname);
+		daos_file_mkdir(&daos_dst_file, ap->src_path, O_RDWR);
+		rc = fs_copy(&daos_src_file, &daos_dst_file,
+				ap->src_path, ap->dst_path);
+		if (rc != 0)
+			D_GOTO(out, rc);
+	} else if (daos_src_file.type == DAOS && daos_dst_file.type == POSIX) {
+		parse_filename(ap->dst_path, &name, &dname);
+		ap->src_path = "/";
+		rc = fs_copy(&daos_src_file, &daos_dst_file,
+				ap->src_path, dname);
+		if (rc != 0)
+			D_GOTO(out, rc);
+	} else if (daos_src_file.type == DAOS && daos_dst_file.type == DAOS) {
+		ap->src_path = "/";
+		ap->dst_path = "/";
+		rc = fs_copy(&daos_src_file, &daos_dst_file,
+				ap->src_path, dname);
```

Comment:
dname is NULL here?

Reply:
@mchaarawi yeah, it won't be used in the DAOS -> DAOS case in fs_copy it is only used when the dest paths need to be constructed for POSIX